### Importing Library:

In [1]:
import pandas as pd

### Loading "Matches" Dataset:

In [2]:
data1 = pd.read_csv("matches.csv")
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
data1.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

### Filtering RR and KXIP matches only:

In [3]:
kxip_rr = data1.loc[((data1['team1'] == 'Kings XI Punjab') & (data1['team2'] == 'Rajasthan Royals') | (data1['team1'] == 'Rajasthan Royals') & (data1['team2'] == 'Kings XI Punjab'))]
kxip_rr.head(10)

id  season        city        date             team1             team2  \
64    65    2008      Jaipur  2008-04-21   Kings XI Punjab  Rajasthan Royals   
96    97    2008  Chandigarh  2008-05-28   Kings XI Punjab  Rajasthan Royals   
129  130    2009   Cape Town  2009-04-26   Kings XI Punjab  Rajasthan Royals   
144  145    2009      Durban  2009-05-05  Rajasthan Royals   Kings XI Punjab   
192  193    2010  Chandigarh  2010-03-24  Rajasthan Royals   Kings XI Punjab   
211  212    2010      Jaipur  2010-04-07   Kings XI Punjab  Rajasthan Royals   
255  256    2011  Chandigarh  2011-04-21   Kings XI Punjab  Rajasthan Royals   
310  311    2012      Jaipur  2012-04-06  Rajasthan Royals   Kings XI Punjab   
353  354    2012  Chandigarh  2012-05-05  Rajasthan Royals   Kings XI Punjab   
397  398    2013      Jaipur  2013-04-14   Kings XI Punjab  Rajasthan Royals   

          toss_winner toss_decision  result  dl_applied            winner  \
64    Kings XI Punjab           bat  normal           0  Rajasthan Royals   
96   Rajasthan Royals         field  normal           0   Kings XI Punjab   
129   Kings XI Punjab           bat  normal           0   Kings XI Punjab   
144   Kings XI Punjab         field  normal           0  Rajasthan Royals   
192   Kings XI Punjab         field  normal           0  Rajasthan Royals   
211   Kings XI Punjab           bat  normal           0  Rajasthan Royals   
255  Rajasthan Royals         field  normal           0   Kings XI Punjab   
310   Kings XI Punjab         field  normal           0  Rajasthan Royals   
353  Rajasthan Royals           bat  normal           0  Rajasthan Royals   
397  Rajasthan Royals         field  normal           0  Rajasthan Royals   

     win_by_runs  win_by_wickets player_of_match  \
64             0               6       SR Watson   
96            41               0        SE Marsh   
129           27               0   KC Sangakkara   
144           78               0        GC Smith   
192           31               0        AC Voges   
211            0               9         MJ Lumb   
255           48               0        SE Marsh   
310           31               0       AM Rahane   
353           43               0       SR Watson   
397            0               6     JP Faulkner   

                                          venue      umpire1        umpire2  \
64                       Sawai Mansingh Stadium    Aleem Dar      RB Tiffin   
96   Punjab Cricket Association Stadium, Mohali     SJ Davis    K Hariharan   
129                                    Newlands    M Erasmus    K Hariharan   
144                                   Kingsmead    SS Hazare      IL Howell   
192  Punjab Cricket Association Stadium, Mohali  BR Doctrove    SK Tarapore   
211                      Sawai Mansingh Stadium       S Ravi    SK Tarapore   
255  Punjab Cricket Association Stadium, Mohali     S Asnani     PR Reiffel   
310                      Sawai Mansingh Stadium    BF Bowden    SK Tarapore   
353  Punjab Cricket Association Stadium, Mohali    JD Cloete     SJA Taufel   
397                      Sawai Mansingh Stadium    Aleem Dar  C Shamshuddin   

    umpire3  
64      NaN  
96      NaN  
129     NaN  
144     NaN  
192     NaN  
211     NaN  
255     NaN  
310     NaN  
353     NaN  
397     NaN

## Question 1: Who will win the match?

#### Solution: Checking counts of winner:

In [6]:
kxip_rr['winner'].value_counts()/kxip_rr.shape[0] * 100

Rajasthan Royals    52.631579
Kings XI Punjab     47.368421
Name: winner, dtype: float64

**Answer:** As per the above analysis, winning percentage of KXIP is **52.63** against RR.

## Question 2: 

#### Solution: Observing runs of KL Rahul in each match:

### Loading "Deliveries" dataset:

In [7]:
data2 = pd.read_csv("deliveries.csv")
data2.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  wide_runs  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0          0         0   
1     2  DA Warner    S Dhawan  TS Mills              0          0         0   
2     3  DA Warner    S Dhawan  TS Mills              0          0         0   
3     4  DA Warner    S Dhawan  TS Mills              0          0         0   
4     5  DA Warner    S Dhawan  TS Mills              0          2         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs player_dismissed dismissal_kind fielder  
0           0              NaN            NaN     NaN  
1           0              NaN            NaN     NaN  
2           4              NaN            NaN     NaN  
3           0              NaN            NaN     NaN  
4           2              NaN            NaN     NaN

In [9]:
rahul = data2.loc[(data2['batsman'] == 'KL Rahul') & ((data2['over'] >= 1) & data2['over'] <= 6)]
rahul_balls = pd.DataFrame(rahul.groupby('match_id')['ball'].count())
rahul_runs = pd.DataFrame(rahul.groupby('match_id')['batsman_runs'].sum())
rahul_data = pd.merge(rahul_balls, rahul_runs, on='match_id')
rahul_data['Strike Rate'] = 100 * rahul_data['batsman_runs']/rahul_data['ball']
rahul_data = rahul_data.rename({'ball' : 'Balls', 'batsman_runs' : 'Runs'}, axis = 1)
rahul_data

Balls  Runs  Strike Rate
match_id                          
401          12    12   100.000000
429           4     8   200.000000
461          19    20   105.263158
466          28    27    96.428571
474          27    25    92.592593
477          40    46   115.000000
481           6     6   100.000000
487          18    18   100.000000
493          13    10    76.923077
500          15    14    93.333333
521           8     5    62.500000
525          32    44   137.500000
529           5     2    40.000000
531          21    24   114.285714
537           3     0     0.000000
541          28    25    89.285714
564          10    17   170.000000
572          26    25    96.153846
590          14    23   164.285714
592          10     7    70.000000
595          36    51   141.666667
603          29    51   175.862069
606          33    52   157.575758
611          35    38   108.571429
615          26    42   161.538462
617          54    68   125.925926
626           7    16   228.571429
632          23    38   165.217391
633           1     0     0.000000
636           9    11   122.222222
7895         16    51   318.750000
7901         30    48   160.000000
7905         22    37   168.181818
7909         21    18    85.714286
7911         28    61   217.857143
7915         15    23   153.333333
7918         26    32   123.076923
7927         20    24   120.000000
7931         56    86   153.571429
7933         70   101   144.285714
7937         31    68   219.354839
7941         15    22   146.666667
7943         64    99   154.687500
7949         12     8    66.666667
11140         4     4   100.000000
11142         5     1    20.000000
11145        58    72   124.137931
11149        11    15   136.363636
11309        47    55   117.021277
11313        53    71   133.962264
11315        66   104   157.575758
11319        15    18   120.000000
11323        47    53   112.765957
11328         9    12   133.333333
11333        29    44   151.724138
11339        60    84   140.000000
11343         7     2    28.571429
11346        36    75   208.333333

## Question 3: 

In [11]:
df = data2.loc[(data2['match_id'] >= 518) & (data2['batting_team'] == 'Delhi Capitals')]
dc_runs = pd.DataFrame(df.groupby('match_id')['total_runs'].sum())
dc_runs = dc_runs.rename({'total_runs' : 'DC Runs'}, axis = 1)
shreyas_runs = data2.loc[(data2['match_id'] >= 518) & (data2['batsman'] == 'SS Iyer')]
shreyas_runs = pd.DataFrame(shreyas_runs.groupby('match_id')['batsman_runs'].sum())
shreyas_runs = shreyas_runs.rename({'batsman_runs' : "Sheyas' Runs"}, axis = 1)
dc_data = pd.merge(dc_runs, shreyas_runs, on = 'match_id')
dc_data["Shreyas' Contribution"] = dc_data["Sheyas' Runs"]/dc_data['DC Runs'] * 100
dc_data

DC Runs  Sheyas' Runs  Shreyas' Contribution
match_id                                              
519           149             7               4.697987
523           184            40              21.739130
528           169             6               3.550296
531           167            60              35.928144
535           146            31              21.232877
539           190            83              43.684211
543            95             0               0.000000
545           158            40              25.316456
548           119            54              45.378151
553           175             9               5.142857
556           152            19              12.500000
561           157             0               0.000000
565           120            70              58.333333
571           187            20              10.695187
578            98             0               0.000000
583           113             3               2.654867
587           192             0               0.000000
593           164            19              11.585366
602           186             0               0.000000
625           121             8               6.611570
7895          172            13               7.558140
7902          197            27              13.705584
7906          136             4               2.941176
7912          186            52              27.956989
7915          148            58              39.189189
7919          223            97              43.497758
7923          206            13               6.310680
7925          203            50              24.630542
7929          166            44              26.506024
7935          190             3               1.578947
7938          188            35              18.617021
7945          171            21              12.280702
7948          186             6               3.225806
11139         219            16               7.305936
11141         154            18              11.688312
11146         199            49              24.623116
11149         163            30              18.404908
11152         134            43              32.089552
11311         164            73              44.512195
11317         183             6               3.278689
11321         164            48              29.268293
11325         136             3               2.205882
11328         177            61              34.463277
11331         194             4               2.061856
11337         193            52              26.943005
11341          99            44              44.444444
11344         126            15              11.904762
11413         170             8               4.705882
11414         155            14               9.032258

## Question 4: 

In [10]:
dot_balls = data2.loc[((data2['batting_team'] == "Kings XI Punjab") & (data2['bowling_team'] == "Kolkata Knight Riders") | (data2['batting_team'] == "Kolkata Knight Riders") & (data2['bowling_team'] == "Kings XI Punjab"))]
dot_balls = dot_balls.loc[dot_balls['total_runs'] ==0]
dot_balls = pd.DataFrame(dot_balls.groupby('match_id')['total_runs'].count()).sort_values(by='total_runs', ascending=False)
dot_balls

total_runs
match_id            
472              100
270               98
514               97
112               92
415               91
81                88
532               88
142               87
196               85
325               85
400               85
11142             81
560               80
11                79
491               76
517               76
48                75
330               75
608               75
589               73
123               72
11343             66
7937              63
7911              60
208               54

### Average wickets:

In [11]:
dot_balls['total_runs'].mean()

80.04

## Question 5: What will be the status of Chris Gayle's batting durin KXIP innings?

#### Solution: Filtering data with CSK and Delhi Matches:

In [22]:
gayle = data2.loc[(data2['batsman'] == 'CH Gayle')]
gayle_balls = pd.DataFrame(gayle.groupby('match_id')['ball'].count())
gayle_runs = pd.DataFrame(gayle.groupby('match_id')['batsman_runs'].sum())
gayle_data = pd.merge(gayle_balls, gayle_runs, on='match_id')
gayle_data['Strike Rate'] = 100 * gayle_data['batsman_runs']/gayle_data['ball']
gayle_data = gayle_data.rename({'ball' : 'Balls', 'batsman_runs' : 'Runs'}, axis = 1)
gayle_data

Balls  Runs  Strike Rate
match_id                          
1            23    32   139.130435
5             9     6    66.666667
12           28    22    78.571429
20           40    77   192.500000
27           18     7    38.888889
30           12     8    66.666667
42            2     0     0.000000
45            1     0     0.000000
55           39    48   123.076923
121          12    10    83.333333
123          29    44   151.724138
126          44    54   122.727273
132          12    12   100.000000
134          38    40   105.263158
138           7     7   100.000000
142          17    17   100.000000
191          65    75   115.384615
196          13    14   107.692308
200          22    30   136.363636
204           7     4    57.142857
208          44    88   200.000000
213          22    40   181.818182
217          17    34   200.000000
222           4     7   175.000000
227           3     0     0.000000
258          57   102   178.947368
263          14    26   185.714286
268          26    49   188.461538
280          49   107   218.367347
283          17    44   258.823529
288          44    70   159.090909
291          12    38   316.666667
296           7     0     0.000000
302          54    75   138.888889
304           9     8    88.888889
306          50    89   178.000000
307           3     0     0.000000
317           8     2    25.000000
320          40    68   170.000000
326           9     8    88.888889
329          49    81   165.306122
333          61    87   142.622951
338           8     4    50.000000
344          59    86   145.762712
350          44    71   161.363636
356          23    26   113.043478
360          59    82   138.983051
362          32    57   178.125000
367           8     6    75.000000
372          62   128   206.451613
376          10    27   270.000000
383          61    92   150.819672
388           6    11   183.333333
392          55    85   154.545455
396          11     4    36.363636
401          12    15   125.000000
407          49    49   100.000000
411          69   175   253.623188
417          20    18    90.000000
420          17    34   200.000000
426          21    21   100.000000
429          56    77   137.500000
432          15    13    86.666667
438           7     4    57.142857
441          37    33    89.189189
448          34    61   179.411765
452          14    28   200.000000
475           7    20   285.714286
481          20    27   135.000000
484          25    38   152.000000
488           7     4    57.142857
492          25    19    76.000000
495          23    22    95.652174
499          52    46    88.461538
503          20    14    70.000000
506           9     6    66.666667
522          57    96   168.421053
525          16    21   131.250000
534          31    10    32.258065
540          17    20   117.647059
543          42    62   147.619048
546           4    10   250.000000
550          12    21   175.000000
557          63   117   185.714286
562          13    13   100.000000
566          15    17   113.333333
568          11    35   318.181818
571           5     1    20.000000
574          28    27    96.428571
575          45    41    91.111111
580           5     1    20.000000
587           3     0     0.000000
606           7     7   100.000000
617           6     5    83.333333
620          13     6    46.153846
624          31    49   158.064516
626          35    73   208.571429
632           4     1    25.000000
633          14     9    64.285714
636          41    76   185.365854
7905         34    64   188.235294
7909         65   111   170.769231
7911         39    63   161.538462
7918         22    23   104.545455
7927         44    58   131.818182
7931         11     8    72.727273
7933          2     2   100.000000
7937         17    21   123.529412
7941         14    18   128.571429
7943         11    18   163.636364
7949          2     0     0.000000
11140        52    91   175.000000
11142        14    2

#### Getting mean of the Chris Gayle's strike rate:

In [23]:
gayle_data['Strike Rate'].mean()

126.91220166371855

**Answer:** Average strike rate of Chris Gayle is **127**.